File: bin/managed.rs
use std::collections::BTreeSet;
use std::fs;
use std::io::{ErrorKind, Write};
use std::path::{Component, Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use super::harness::{Harness, SyncEnv};
use super::{err, rm_entry};

#[derive(Clone, Debug)]
pub(super) struct ManagedSyncPlan {
    harnesses: Vec<ManagedHarnessPlan>,
}

#[derive(Clone, Debug)]
struct ManagedHarnessPlan {
    state_path: PathBuf,
    cleanup_paths: Vec<PathBuf>,
    current_entry_names: Vec<String>,
}

pub(super) fn plan_managed_entries(sync_env: &SyncEnv) -> Result<ManagedSyncPlan, String> {
    let harnesses = sync_env
        .harnesses
        .iter()
        .map(|harness| plan_harness(sync_env, harness))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(ManagedSyncPlan { harnesses })
}

pub(super) fn clean_managed_entries(plan: &ManagedSyncPlan) -> bool {
    let mut success = true;
    for harness in &plan.harnesses {
        for path in &harness.cleanup_paths {
            if let Err(error) = rm_entry(path) {
                err(&format!("cleanup failed: {} ({error})", path.display()));
                success = false;
            }
        }
    }
    success
}

pub(super) fn record_managed_entries(plan: &ManagedSyncPlan) -> bool {
    let mut success = true;
    for harness in &plan.harnesses {
        if let Err(message) =
            write_recorded_entry_names(&harness.state_path, &harness.current_entry_names)
        {
            err(&format!("managed state write failed: {message}"));
            success = false;
        }
    }
    success
}

pub(super) fn asset_dir_names(path: &Path) -> Result<Vec<String>, String> {
    dir_entry_names(path, true)
}

pub(super) fn top_level_entry_names(path: &Path) -> Result<Vec<String>, String> {
    dir_entry_names(path, false)
}

fn dir_entry_names(path: &Path, dirs_only: bool) -> Result<Vec<String>, String> {
    if !path.is_dir() {
        return Ok(Vec::new());
    }

    let entries =
        fs::read_dir(path).map_err(|error| format!("read {} ({error})", path.display()))?;
    let mut names = Vec::new();
    for entry_result in entries {
        let entry = entry_result.map_err(|error| format!("read {} ({error})", path.display()))?;
        let file_type = entry
            .file_type()
            .map_err(|error| format!("stat {} ({error})", entry.path().display()))?;
        if dirs_only && !file_type.is_dir() {
            continue;
        }

        let name = entry.file_name();
        let name = name
            .to_str()
            .ok_or_else(|| format!("invalid UTF-8 path: {}", entry.path().display()))?;
        names.push(name.to_string());
    }
    names.sort();
    Ok(names)
}

fn plan_harness(sync_env: &SyncEnv, harness: &Harness) -> Result<ManagedHarnessPlan, String> {
    let current_entry_names = current_managed_entry_names(sync_env, harness)?;
    let state_path = harness.managed_state_path(&sync_env.managed_state_home);
    let mut cleanup_entry_names = BTreeSet::new();
    cleanup_entry_names.extend(current_entry_names.iter().cloned());
    cleanup_entry_names.extend(
        harness
            .compat_managed_entries
            .iter()
            .map(|entry| (*entry).to_string()),
    );
    cleanup_entry_names.extend(load_recorded_entry_names(&state_path)?);

    let harness_root = harness.root();
    Ok(ManagedHarnessPlan {
        state_path,
        cleanup_paths: cleanup_entry_names
            .into_iter()
            .filter_map(|entry| cleanup_path(&harness_root, &entry))
            .collect(),
        current_entry_names,
    })
}

fn current_managed_entry_names(
    sync_env: &SyncEnv,
    harness: &Harness,
) -> Result<Vec<String>, String> {
    let mut names = BTreeSet::new();
    names.insert(harness.instruction_file_name().to_string());

    for entry_name in top_level_entry_names(&harness.source_root(&sync_env.tools_home))? {
        names.insert(entry_name);
    }

    for asset_name in asset_dir_names(&sync_env.assets_home)? {
        names.insert(harness.rename_asset(&asset_name));
    }

    Ok(names.into_iter().collect())
}

fn load_recorded_entry_names(path: &Path) -> Result<Vec<String>, String> {
    let content = match fs::read_to_string(path) {
        Ok(content) => content,
        Err(error) if error.kind() == ErrorKind::NotFound => return Ok(Vec::new()),
        Err(error) => return Err(format!("read {} ({error})", path.display())),
    };

    let recorded: Vec<String> = match serde_json::from_str(&content) {
        Ok(recorded) => recorded,
        Err(error) => {
            warn(&format!(
                "managed state parse failed, ignoring {} ({error})",
                path.display()
            ));
            return Ok(Vec::new());
        }
    };

    let mut safe_names = BTreeSet::new();
    for entry_name in recorded {
        match safe_top_level_entry_name(&entry_name) {
            Some(name) => {
                safe_names.insert(name);
            }
            None => warn(&format!(
                "ignoring unsafe managed entry {:?} in {}",
                entry_name,
                path.display()
            )),
        }
    }
    Ok(safe_names.into_iter().collect())
}

fn write_recorded_entry_names(path: &Path, entry_names: &[String]) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .map_err(|error| format!("create {} ({error})", parent.display()))?;
    }

    let content = format!(
        "{}\n",
        serde_json::to_string_pretty(entry_names)
            .map_err(|error| format!("serialize {} ({error})", path.display()))?
    );
    let (temp_path, mut temp_file) = create_temp_state_file(path)?;

    let result = (|| {
        temp_file
            .write_all(content.as_bytes())
            .map_err(|error| format!("write {} ({error})", temp_path.display()))?;
        temp_file
            .sync_all()
            .map_err(|error| format!("sync {} ({error})", temp_path.display()))?;
        drop(temp_file);
        fs::rename(&temp_path, path)
            .map_err(|error| format!("replace {} ({error})", path.display()))
    })();

    if result.is_err() {
        let _ = fs::remove_file(&temp_path);
    }

    result
}

fn cleanup_path(root: &Path, entry_name: &str) -> Option<PathBuf> {
    safe_top_level_entry_name(entry_name).map(|name| root.join(name))
}

fn safe_top_level_entry_name(entry_name: &str) -> Option<String> {
    let mut components = Path::new(entry_name).components();
    match (components.next(), components.next()) {
        (Some(Component::Normal(name)), None) => name.to_str().map(ToOwned::to_owned),
        _ => None,
    }
}

fn create_temp_state_file(path: &Path) -> Result<(PathBuf, fs::File), String> {
    let file_name = path
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or("managed-state.json");
    let nonce = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO)
        .as_nanos();

    for attempt in 0..16_u32 {
        let temp_path = path.with_file_name(format!(
            ".{file_name}.{}.{}.tmp",
            std::process::id(),
            nonce + u128::from(attempt)
        ));
        match fs::OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(&temp_path)
        {
            Ok(file) => return Ok((temp_path, file)),
            Err(error) if error.kind() == ErrorKind::AlreadyExists => continue,
            Err(error) => return Err(format!("create {} ({error})", temp_path.display())),
        }
    }

    Err(format!(
        "create temporary managed state near {} (name collision)",
        path.display()
    ))
}

fn warn(message: &str) {
    eprintln!("sync: warning: {message}");
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use super::super::harness::HarnessId;
    use tempfile::TempDir;

    use super::*;

    fn write_file(path: &Path, content: &str) {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).expect("create parent dirs");
        }
        fs::write(path, content).expect("write file")
    }

    #[test]
    fn invalid_recorded_entries_are_ignored_and_do_not_escape_harness_root() {
        let temp = TempDir::new().expect("tempdir");
        let sync_env = SyncEnv::from_home(temp.path().to_path_buf(), Duration::from_secs(1));
        let harness = sync_env.harness(HarnessId::Codex).expect("codex harness");
        let state_path = harness.managed_state_path(&sync_env.managed_state_home);
        write_file(
            &state_path,
            r#"[
  "good.txt",
  "..",
  "/tmp/escape",
  "nested/path",
  "../outside",
  "good.txt"
]"#,
        );

        let plan = plan_harness(&sync_env, harness).expect("plan harness");

        assert_eq!(
            plan.cleanup_paths,
            vec![
                harness.root().join("AGENTS.md"),
                harness.root().join("good.txt")
            ]
        );
    }

    #[test]
    fn malformed_recorded_state_is_recoverable() {
        let temp = TempDir::new().expect("tempdir");
        let sync_env = SyncEnv::from_home(temp.path().to_path_buf(), Duration::from_secs(1));
        let harness = sync_env.harness(HarnessId::Codex).expect("codex harness");
        let state_path = harness.managed_state_path(&sync_env.managed_state_home);
        write_file(&state_path, "{not valid json");

        assert_eq!(
            load_recorded_entry_names(&state_path).expect("recover load"),
            Vec::<String>::new()
        );
        let plan = plan_harness(&sync_env, harness).expect("recover plan");
        assert_eq!(plan.cleanup_paths, vec![harness.root().join("AGENTS.md")]);
    }

    #[test]
    fn recorded_state_write_persists_expected_json() {
        let temp = TempDir::new().expect("tempdir");
        let path = temp.path().join("state").join("codex.json");

        write_recorded_entry_names(&path, &["alpha".to_string(), "beta".to_string()])
            .expect("write state");

        assert_eq!(
            fs::read_to_string(&path).expect("read state"),
            "[\n  \"alpha\",\n  \"beta\"\n]\n"
        );
    }
}

File: bin/managed.ts
import fs from "node:fs";
import { dirname, join, posix } from "node:path";

import { Harness, SyncEnv } from "./harness.ts";
import { err, panicMessage, rmEntry, warn } from "./lib.ts";

export interface ManagedSyncPlan {
  harnesses: ManagedHarnessPlan[];
}

export interface ManagedHarnessPlan {
  statePath: string;
  cleanupPaths: string[];
  currentEntryNames: string[];
}

export function planManagedEntries(syncEnv: SyncEnv): ManagedSyncPlan {
  return {
    harnesses: syncEnv.harnesses.map((harness) => planHarness(syncEnv, harness)),
  };
}

export function cleanManagedEntries(plan: ManagedSyncPlan): boolean {
  let success = true;
  for (const harness of plan.harnesses) {
    for (const path of harness.cleanupPaths) {
      try {
        rmEntry(path);
      } catch (error) {
        err(`cleanup failed: ${path} (${panicMessage(error)})`);
        success = false;
      }
    }
  }
  return success;
}

export function recordManagedEntries(plan: ManagedSyncPlan): boolean {
  let success = true;
  for (const harness of plan.harnesses) {
    try {
      writeRecordedEntryNames(harness.statePath, harness.currentEntryNames);
    } catch (error) {
      err(`managed state write failed: ${panicMessage(error)}`);
      success = false;
    }
  }
  return success;
}

export function assetDirNames(path: string): string[] {
  return dirEntryNames(path, true);
}

export function topLevelEntryNames(path: string): string[] {
  return dirEntryNames(path, false);
}

export function loadRecordedEntryNames(path: string): string[] {
  let content: string;
  try {
    content = fs.readFileSync(path, "utf8");
  } catch (error) {
    if (isNotFound(error)) {
      return [];
    }
    throw new Error(`read ${path} (${panicMessage(error)})`);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(content);
  } catch (error) {
    warn(`managed state parse failed, ignoring ${path} (${panicMessage(error)})`);
    return [];
  }

  if (!Array.isArray(parsed)) {
    warn(`managed state parse failed, ignoring ${path} (not an array)`);
    return [];
  }
  if (!parsed.every((entryName) => typeof entryName === "string")) {
    warn(`managed state parse failed, ignoring ${path} (entries must be strings)`);
    return [];
  }

  const safeNames = new Set<string>();
  for (const entryName of parsed) {
    const safeName = safeTopLevelEntryName(entryName);
    if (safeName) {
      safeNames.add(safeName);
    } else {
      warn(`ignoring unsafe managed entry ${JSON.stringify(entryName)} in ${path}`);
    }
  }
  return [...safeNames].sort();
}

export function writeRecordedEntryNames(path: string, entryNames: string[]): void {
  const parent = dirname(path);
  try {
    fs.mkdirSync(parent, { recursive: true });
  } catch (error) {
    throw new Error(`create ${parent} (${panicMessage(error)})`);
  }

  let content: string;
  try {
    content = `${JSON.stringify(entryNames, null, 2)}\n`;
  } catch (error) {
    throw new Error(`serialize ${path} (${panicMessage(error)})`);
  }
  const { tempPath, fd } = createTempStateFile(path);
  try {
    try {
      fs.writeFileSync(fd, content, "utf8");
    } catch (error) {
      throw new Error(`write ${tempPath} (${panicMessage(error)})`);
    }
    try {
      fs.fsyncSync(fd);
    } catch (error) {
      throw new Error(`sync ${tempPath} (${panicMessage(error)})`);
    }
    try {
      fs.closeSync(fd);
    } catch {
      // ignore
    }
    try {
      fs.renameSync(tempPath, path);
    } catch (error) {
      throw new Error(`replace ${path} (${panicMessage(error)})`);
    }
  } catch (error) {
    try {
      fs.closeSync(fd);
    } catch {
      // ignore
    }
    try {
      fs.rmSync(tempPath, { force: true });
    } catch {
      // ignore
    }
    throw error;
  }
}

function dirEntryNames(path: string, dirsOnly: boolean): string[] {
  if (!isDirectory(path)) {
    return [];
  }

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(path, { withFileTypes: true });
  } catch (error) {
    throw new Error(`read ${path} (${panicMessage(error)})`);
  }
  const names: string[] = [];
  for (const entry of entries) {
    if (dirsOnly && !entry.isDirectory()) {
      continue;
    }
    names.push(entry.name);
  }
  names.sort();
  return names;
}

function planHarness(syncEnv: SyncEnv, harness: Harness): ManagedHarnessPlan {
  const currentEntryNames = currentManagedEntryNames(syncEnv, harness);
  const statePath = harness.managedStatePath(syncEnv.managedStateHome);
  const cleanupEntryNames = new Set<string>();
  for (const name of currentEntryNames) {
    cleanupEntryNames.add(name);
  }
  for (const name of harness.compatManagedEntries) {
    cleanupEntryNames.add(name);
  }
  for (const name of loadRecordedEntryNames(statePath)) {
    cleanupEntryNames.add(name);
  }

  const harnessRoot = harness.root();
  return {
    statePath,
    cleanupPaths: [...cleanupEntryNames]
      .map((entry) => cleanupPath(harnessRoot, entry))
      .filter((entry): entry is string => entry !== null),
    currentEntryNames,
  };
}

function currentManagedEntryNames(syncEnv: SyncEnv, harness: Harness): string[] {
  const names = new Set<string>();
  names.add(harness.instructionFileName());
  for (const entryName of topLevelEntryNames(harness.sourceRoot(syncEnv.toolsHome))) {
    names.add(entryName);
  }
  for (const assetName of assetDirNames(syncEnv.assetsHome)) {
    names.add(harness.renameAsset(assetName));
  }
  return [...names].sort();
}

function cleanupPath(root: string, entryName: string): string | null {
  const safeName = safeTopLevelEntryName(entryName);
  return safeName ? join(root, safeName) : null;
}

function safeTopLevelEntryName(entryName: string): string | null {
  if (entryName.length === 0) {
    return null;
  }
  if (posix.isAbsolute(entryName)) {
    return null;
  }
  if (entryName.includes("/")) {
    return null;
  }
  if (entryName === "." || entryName === "..") {
    return null;
  }
  return entryName;
}

function createTempStateFile(path: string): { tempPath: string; fd: number } {
  const baseName = path.split("/").pop() ?? "managed-state.json";
  const nonce = Date.now().toString(16);
  for (let attempt = 0; attempt < 16; attempt += 1) {
    const tempPath = join(dirname(path), `.${baseName}.${process.pid}.${nonce}-${attempt}.tmp`);
    try {
      const fd = fs.openSync(tempPath, "wx");
      return { tempPath, fd };
    } catch (error) {
      if (!isAlreadyExists(error)) {
        throw new Error(`create ${tempPath} (${panicMessage(error)})`);
      }
    }
  }

  throw new Error(`create temporary managed state near ${path} (name collision)`);
}

function isDirectory(path: string): boolean {
  try {
    return fs.statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function isNotFound(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && (error as { code?: string }).code === "ENOENT";
}

function isAlreadyExists(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && (error as { code?: string }).code === "EEXIST";
}
